### PR TITLE
Add undo&redo to the Form editor [MAILPOET-3085]

### PR DESCRIPTION
**Pull request by lysyjan**
[MAILPOET-3085]

[MAILPOET-3085]: https://mailpoet.atlassian.net/browse/MAILPOET-3085

### CODE REVIEW

**Comment by lysyjan** [comment]
Hey, please look at this PR. I'm open to all discussions.
I had troubles with [Selection.jsx](https://github.com/mailpoet/mailpoet/blob/master/assets/js/src/form_editor/components/form_settings/selection.jsx) and I didn't find another way.
Some issues were with displaying when I used undo&redo.

File: assets/js/src/form_editor/store/actions.ts [hunk 1]
```diff
@@ -33,15 +33,21 @@ export function* toggleFullscreen(toggleTo) {
   };
 }
 
-export function changeFormBlocks(blocks) {
-  return {
+export function* changeFormBlocks(blocks) {
+  yield {
+    type: 'CREATE_HISTORY_RECORD',
```
Comment (pavel-mailpoet) [inline assets/js/src/form_editor/store/actions.ts:38]: Why do we need a new action? This is called when other actions are also triggered. I don't see why is this necessary. We can have this in the reducer, I think that would be nicer and keep the actions cleaner

File: assets/js/src/form_editor/store/actions.ts [hunk 2]
```diff
@@ -295,3 +307,23 @@ export function* applyStylesToAllTextInputs(styles) {
     styles,
   };
 }
+
+export function createHistoryRecord() {
```
Comment (pavel-mailpoet) [inline assets/js/src/form_editor/store/actions.ts:311]: I don't see this being used, is it?

File: assets/js/src/form_editor/store/actions.ts [hunk 3]
```diff
@@ -295,3 +307,23 @@ export function* applyStylesToAllTextInputs(styles) {
...
+  return {
+    type: 'CREATE_HISTORY_RECORD',
+  };
+}
+
+export function* historyMove(action) {
+  const editorHistoryOffset = select('mailpoet-form-editor').getEditorHistoryOffset();
+  if (editorHistoryOffset === 0) {
```
Comment (pavel-mailpoet) [inline assets/js/src/form_editor/store/actions.ts:319]: I am not sure I understand. Why do we want to create a new action here? I mean we are at the beginning of the history, right? So I think nothing should happen when the user goes back. So I don't understand why is here the `CREATE_HISTORY_RECORD` triggered.

File: assets/js/src/form_editor/components/history_redo.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import { redo as redoIcon } from '@wordpress/icons';
+
+function HistoryRedo(props) {
+  const hasRedo = useSelect(
+    (select) => select('mailpoet-form-editor').hasEditorRedo(),
+    []
+  );
+  const { historyMove } = useDispatch('mailpoet-form-editor');
+  const redoAction = () => {
+    historyMove('redo');
```
Comment (pavel-mailpoet) [inline assets/js/src/form_editor/components/history_redo.tsx:14]: You can convince me otherwise, but I think it is nicer to have two different functions for undo and redo rather than using a string argument. We wouldn't have to check the value of the argument and it reads a bit better I think.

File: assets/js/src/form_editor/store/selectors.jsx [hunk 5]
```diff
@@ -187,4 +187,18 @@ export default {
       return result;
     }, null);
   },
+  getEditorHistoryOffset(state) {
+    return state.editorHistoryOffset;
+  },
+  hasEditorUndo(state) {
```
Comment (pavel-mailpoet) [inline assets/js/src/form_editor/store/selectors.jsx:193]: Do you think we should write a test for this function?

File: assets/js/src/form_editor/store/selectors.jsx [hunk 6]
```diff
@@ -187,4 +187,18 @@ export default {
...
+    let offset = state.editorHistoryOffset;
+    if (state.editorHistory.length > 1) {
```
Comment (pavel-mailpoet) [inline assets/js/src/form_editor/store/selectors.jsx:195]: Do you think this would deserve a comment? It took me awhile to understand why do we need to add 1.

File: assets/js/src/form_editor/components/editor.jsx [hunk 7]
```diff
@@ -44,6 +44,11 @@ export default () => {
     (sel) => sel('mailpoet-form-editor').getSidebarOpened(),
     []
   );
+  const formBlocks = useSelect(
+    (sel) => sel('mailpoet-form-editor').getFormBlocks(),
```
Comment (pavel-mailpoet) [inline assets/js/src/form_editor/components/editor.jsx:48]: Nothing wrong here, I just wonder. Why did you have to make this change? What is wrong with `select('mailpoet-form-editor').getFormBlocks()`?

File: assets/js/src/form_editor/components/form_settings/selection.jsx [hunk 8]
```diff
@@ -26,6 +26,17 @@ class Selection extends React.Component {
         .trigger('change');
     }
 
+    if (
```
Comment (pavel-mailpoet) [inline assets/js/src/form_editor/components/form_settings/selection.jsx:29]: I think we need a comment or something to understand what is going on. The commit message is not very helpful here and if we come back in two years we will wonder why is this code here.

File: assets/js/src/form_editor/store/reducers/history_record.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
+import { isEqual } from 'lodash';
+
+export const createHistoryRecord = (state) => {
+  let editorHistory: any[] = state.editorHistory;
```
Comment (pavel-mailpoet) [inline assets/js/src/form_editor/store/reducers/history_record.ts:4]: It is great that you are using types here. But please don't use `any`, in my branch I updated TS to disallow this. The build would fail.
You know the structure, right? So it should be easier to create types for this.

File: assets/js/src/form_editor/store/reducers/history_record.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
+import { isEqual } from 'lodash';
+
+export const createHistoryRecord = (state) => {
```
Comment (pavel-mailpoet) [inline assets/js/src/form_editor/store/reducers/history_record.ts:3]: I think this whole file deserves a test, what do you think?

File: assets/js/src/form_editor/store/reducers/history_record.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+  let editorHistory: any[] = state.editorHistory;
+  let editorHistoryOffset = state.editorHistoryOffset;
+  const lastHistoryRecord = editorHistory[editorHistory.length - 1];
+
+  let newHistoryRecord;
+  const time = Date.now();
+
+  if (
+    lastHistoryRecord !== undefined
+    && (time - lastHistoryRecord.time) < 1000
```
Comment (pavel-mailpoet) [inline assets/js/src/form_editor/store/reducers/history_record.ts:13]: Would it be better if we used a constants instead of using a magical numbers 1000 or 100 here?

File: assets/js/src/form_editor/store/reducers/history_record.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    && editorHistory.length > 1
+  ) {
```
Comment (pavel-mailpoet) [inline assets/js/src/form_editor/store/reducers/history_record.ts:15]: I think this function need either split into a couple of smaller ones and/or some comments why do we need all of the logic here.

**Review by pavel-mailpoet — COMMENTED** [review]
Hey, good job doing this. It looks great. I just have a couple of nitpicks. I am open to discussion about all of them. I think we need a couple of tests though, unit tests and maybe even an acceptance test for the new functionality.

File: assets/js/src/form_editor/store/actions.ts (hunk 3 above)
Comment (lysyjan) [reply assets/js/src/form_editor/store/actions.ts:319]: This is an excellent question. I will try to explain it by example. When you make one change then you will have one history record. Right? And when you return back then you will have restored the same record that you have in history. And my question is. What do you restore when you will want to move forward? Without creating a record when you move from zero you haven't the last state. Do you agree?

File: assets/js/src/form_editor/store/reducers/history_record.ts (hunk 11 above)
Comment (lysyjan) [reply assets/js/src/form_editor/store/reducers/history_record.ts:13]: Yes, you are right. I'm not used to write constants in TS. I'll be better for next time.

File: assets/js/src/form_editor/store/actions.ts (hunk 3 above)
Comment (pavel-mailpoet) [reply assets/js/src/form_editor/store/actions.ts:319]: I am not sure I follow your logic, but I reckon that still needs an explanation in the code then:

> When you make one change then you will have one history record. Right?
right, I follow that logic
> And when you return back then you will have restored the same record that you have in history
I think I follow, but you mean do undo and redo, correct?
>  What do you restore when you will want to move forward?
I don't know, why do you need to restore anything, you are at the end of the queue, you are not? Why then restore something if you are at the end.

Comment (lysyjan) [reply assets/js/src/form_editor/store/actions.ts:319]: > When you make one change then you will have one history record. Right?
You make some change in form. For example, move up with a button.
> And when you return back then you will have restored the same record that you have in history
I mean **undo** 🔙.
> What do you restore when you will want to move forward?
When you would have only one record in history, and after **undo** you want to do **redo**, you will have nothing to restore. This is the reason why I create new history record before **undo**.

File: assets/js/src/form_editor/components/history_redo.tsx (hunk 4 above)
Comment (lysyjan) [reply assets/js/src/form_editor/components/history_redo.tsx:14]: The functions would be almost the same instead of two lines. I don't guess that it's the right way.
Maybe something like a private function and two public functions?

Comment (pavel-mailpoet) [reply assets/js/src/form_editor/components/history_redo.tsx:14]: Yeah, one private with two public functions sound like a great plan.

File: assets/js/src/form_editor/components/editor.jsx (hunk 7 above)
Comment (lysyjan) [reply assets/js/src/form_editor/components/editor.jsx:48]: Form Block didn't refresh when I clicked on undo without useSelect.

File: assets/js/src/form_editor/store/actions.ts (hunk 2 above)
Comment (lysyjan) [reply assets/js/src/form_editor/store/actions.ts:311]: Right, it's redundant.

File: assets/js/src/form_editor/store/reducers/history_record.ts (hunk 9 above)
Comment (lysyjan) [reply assets/js/src/form_editor/store/reducers/history_record.ts:4]: Right, it's a nice place for type.

File: assets/js/src/form_editor/store/reducers/history_record.ts (hunk 10 above)
Comment (lysyjan) [reply assets/js/src/form_editor/store/reducers/history_record.ts:3]: 👍🏻 Absolutely yes.

File: assets/js/src/form_editor/store/actions.ts (hunk 1 above)
Comment (lysyjan) [reply assets/js/src/form_editor/store/actions.ts:38]: It isn't necessary. I preferred this solution. I think that this functionality will be in reducers a little bit hidden. But if you want to move it I have no problem with it.

Comment (costasovo) [reply assets/js/src/form_editor/store/actions.ts:38]: I think putting it in reducer would be nicer. We have one big reducer for all actions, but we can create another reducer only for undo/redo it would listen to all actions that need creating history. We can use [combineReducers](https://developer.wordpress.org/block-editor/packages/packages-data/#combineReducers) to combine the reducer we have now and this undo-redo. I think this is the recommended approach.

File: tests/acceptance/FormEditorUndoAndRedoCest.php [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    $formName = 'My fancy form';
+    $form = new Form();
+    $form->withName($formName)->withSegments([$segment])->create();
+    $i->wantTo('Add first and last name to the editor');
+    $i->login();
+    $i->amOnMailPoetPage('Forms');
+    $i->waitForText($formName);
+    $i->clickItemRowActionByItemName($formName, 'Edit');
+    $i->waitForElement('[data-automation-id="form_title_input"]');
+    // Add First name block
```
Comment (pavel-mailpoet) [inline tests/acceptance/FormEditorUndoAndRedoCest.php:22]: Instead of using comments in the Acceptance tests, you should use `$i->wantTo()`. It has the same meaning and can be displayed in the output if using verbose output.

**Review by pavel-mailpoet — DISMISSED** [review]

**Review by veljkho — CHANGES_REQUESTED** [review]
Everything's good except keyboard shortcuts. It just won't work with the keys.

File: assets/js/src/form_editor/components/history_redo.tsx [hunk 14]
```diff
@@ -3,21 +3,47 @@ import { __ } from '@wordpress/i18n';
...
+    (event) => {
+      redoAction();
+      event.preventDefault();
+    },
+  );
+
+  registerShortcut({
+    name: 'mailpoet-form-editor/redo',
+    category: 'block',
+    description: __('Redo your last undo.'),
```
Comment (pavel-mailpoet) [inline assets/js/src/form_editor/components/history_redo.tsx:34]: This is using the gutenberg default translation function. Is that intentional?

Comment (lysyjan) [reply assets/js/src/form_editor/components/history_redo.tsx:34]: Yes, both translations are from Gutenberg.

**Review by pavel-mailpoet — APPROVED** [review]

**Review by veljkho — APPROVED** [review]
